ggsql announcement (#170)
posit-dev/open-source-website: content/blog/ggsql_alpha_release/index.qmd

![](epic_handshake.png)

From the above, slightly hyperbolic, overview it is clear that both SQL and the grammar of graphics share a lot of commonality in their approach to their respective domains. Together they can offer a very powerful and natural solution to the full pipeline from raw data to final visualization.

## Meet ggsql
That was a lot of words about the why. Hopefully you come away from it all with a sense of this whole thing being justified, ready to learn more about it. If so, read on!

### A (simple) example
Let's start with some code even though we haven't introduced the syntax yet. I promise we will go through it below. The example is an adaptation of a visualization created by [Jack Davison](https://jack-davison.github.io) for TidyTuesday.

> ```{ggsql}

Reviewer: I'm not sure if we could do this right now, but this would be a good spot where can use hover/select code annotation: https://quarto.org/docs/authoring/code-annotation.html.

      'John Glenn was 77\non his last mission -\nthe oldest person to\ntravel in space!'
    ],
    hjust => 'left',
    vjust => 'top',
    offset => [10, 0]
SCALE fill TO accent
LABEL
  title => 'How old are astronauts on their most recent mission?',
  subtitle => 'Age of astronauts when they were selected and when they were sent on their mission',
  x => 'Age of astronaut (years)',
  fill => null
```

That was a lot of code, but on the flip-side it allows us to cover a lot of the most important aspects of the syntax with one example.

At the topmost level there are two parts to an SQL query: The SQL query, and the visualization query. The SQL query is anything from the beginning to the `VISUALIZE` clause. It is your standard SQL, and it accepts anything your backend accepts (in this blog post we use a DuckDB backend). The result of the query is funnelled directly into the visualization rather than being returned as a table like you'd normally expect. It should be noted that the SQL query part is optional. If your data is already in the right shape for plotting you can skip it and instead name the source directly in the `VISUALIZE` clause:

```ggsql

> VISUALIZE year_of_selection AS x, year_of_mission AS y FROM 'astronauts.parquet'

Reviewer 2: Is it worth adding `DRAW point` here? I don't think it makes it any less clear and you get a pretty plot to look at.

Reviewer 2: Oh yes I see it might distract from the above plot.

ggsql is a blank slate. Not only in the sense that we are building it from the ground up, but also in that it is built for an environment with no established expectations for a visualization tool. I cannot stress how liberating and invigorating this has felt, and I am positive that this shines through in how ggsql feels to the user.

## Meet ggsql
That was a lot of words about the why. Hopefully you come away from it all with a sense of this whole thing being justified, ready to learn more about it. If so, read on!

### A (simple) example
Let's start with some code even though we haven't introduced the syntax yet. I promise we will go through it below. The example is an adaptation of a visualization created by [Jack Davison](https://jack-davison.github.io) for TidyTuesday.

> ```{ggsql}

Reviewer 2: Any chance at all of somehow getting this syntax highlighted in the final rendered post?

Author: no... sadly not

    PARTITION BY name
    ORDER BY mission_number DESC
  ) = 1
)
SELECT                                          
  *, 
  year_of_selection - year_of_birth AS age,
  'Age at selection' AS category
FROM astronauts
UNION ALL
SELECT 
  *, 
  year_of_mission - year_of_birth AS age,
  'Age at mission' AS category
FROM astronauts

VISUALIZE age AS x, category AS fill
DRAW histogram
  SETTING binwidth => 1, position => 'identity'
PLACE rule
  SETTING x => [34, 44], linetype => 'dotted'
PLACE text
  SETTING
    x => [34, 44, 60],
    y => [66, 49, 20],
    label => [
      'Mean age at selection = 34',
      'Mean age at mission = 44',
      'John Glenn was 77\non his last mission -\nthe oldest person to\ntravel in space!'
    ],
    hjust => 'left',
    vjust => 'top',
    offset => [10, 0]
SCALE fill TO accent
LABEL
  title => 'How old are astronauts on their most recent mission?',
  subtitle => 'Age of astronauts when they were selected and when they were sent on their mission',
  x => 'Age of astronaut (years)',
  fill => null
```

That was a lot of code, but on the flip-side it allows us to cover a lot of the most important aspects of the syntax with one example.

> At the topmost level there are two parts to an SQL query: The SQL query, and the visualization query. The SQL query is anything from the beginning to the `VISUALIZE` clause. It is your standard SQL, and it accepts anything your backend accepts (in this blog post we use a DuckDB backend). The result of the query is funnelled directly into the visualization rather than being returned as a table like you'd normally expect. It should be noted that the SQL query part is optional. If your data is already in the right shape for plotting you can skip it and instead name the source directly in the `VISUALIZE` clause:

Reviewer: I find the order here uncomfortable.
I think the flow would be better if from `It should be noted that ...` to the code block is moved down a paragraph.
The 'Since the point of this post ...' paragraph is all about how we move on from SQL to visualise and a natural end to that paragraph is 'You don't need the SQL part of all'.

As an example, let's change the above plot to instead show the same relationship as a jittered scatterplot.

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter'
```

Or perhaps the jitter follows the distribution of the data so it doubles as a violin plot:

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter', distribution => 'density'
```

As you can see the syntax and composable nature makes visualization iteration very ergonomic, something that is extremely valuable in both explorative analyses and visualization design.

## Behind the curtains
With a passing understanding of how to use ggsql, let's discuss what goes on behind the scenes. At its core ggsql is a modular Rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.

#### The Reader

> This module is responsible for communicating with your data backend, i.e. your SQL database. While this makes it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org). DuckDB allows you to ingest e.g. parquet and csv files directly (no installation necessary) as we have done above, as well as talk to remote SQL databases through its rich (extension ecosystem)[https://duckdb.org/docs/current/core_extensions/overview].

Reviewer: ```suggestion
This module is responsible for communicating with your data backend, i.e. your SQL database. While this makes it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org). DuckDB allows you to ingest e.g. parquet and csv files directly (no installation necessary) as we have done above, as well as talk to remote SQL databases through its rich [extension ecosystem](https://duckdb.org/docs/current/core_extensions/overview).
```

#### The Plot

> The Reader module communicates with the Plot module in two ways. It tells the Plot module about it's particular flavor of SQL and allows it to execute queries on it, returning the resulting data. On the other hand, the Plot module is responsible for reading the visualize query and parse it into a plot representation that can be handed off to the Writer. Part of this is to construct a single query for each layer in the visualization (each `DRAW` query) that returns the data needed for that layer. All computations necessary for the layer is part of this query and are thus handled by the backend. For instance, the query for the boxplot layer includes logic for calculating quantiles etc, all happening in the database without the need for materializing the raw data.

Reviewer: ```suggestion
The Reader module communicates with the Plot module in two ways. It tells the Plot module about it's particular flavor of SQL and allows it to execute queries on it, returning the resulting data. On the other hand, the Plot module is responsible for reading the visualize query and parse it into a plot representation that can be handed off to the Writer. Part of this is to construct a single query for each layer in the visualization (each `DRAW` query) that returns the data needed for that layer. All computations necessary for the layer are part of this query and are thus handled by the backend. For instance: the query for the boxplot layer includes logic for calculating quantiles which all happens in the database without the need for materializing the raw data.
```

#### The Writer

> This module is responsible for the output. It gets the structured representation of the visualization from the Plot module and converts it into the actual visual representation of it. Currently we only have a [Vegalite](https://vega.github.io/vega-lite/) writer but we are planning to expand this very soon to shed our reliance on Vegalite idioms.

Reviewer: ```suggestion
This module is responsible for the output. It gets the structured representation of the visualization from the Plot module and converts it into the actual visual representation of it. Currently we only have a [Vega-Lite](https://vega.github.io/vega-lite/) writer but we are planning to expand this very soon to shed our reliance on Vega-Lite idioms.
```
This is just how they spell it themselves

After the `DRAW` and `PLACE` clauses we have a `SCALE` clause. This clause controls how data values are translated into values that are meaningful for the aesthetic. In our case, the `category` column holds the strings "Age at mission" and "Age at selection" which doesn't in itself translate to a color value. The clause `SCALE fill TO accent` tells ggsql to use the "accent" color palette when converting the values mapped to fill to actual colors. Scales can be used for much more, like applying transformations to continuous data, defining break points, and setting specific scale types (like ordinal or binned).

The last clause in our visual query is `LABEL` which allows us to add or modify various text labels like title, subtitle, and axis and legend titles. 

### Stepping back

> That was a mouthful. But there are two very silvery linings to it all:

Reviewer 3: IMO it would be better to step people up to the complex example. 

If you wanted to keep a complex example first, I'd suggest you just show it, then explain the syntax with simple plots, then come back to an explanation of the complex plot.

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW boxplot
```

That's much shorter but still, if you are coming from a different plotting system you may even think this is overly verbose (e.g. compared to something like `boxplot(astronauts.sex, astronauts.year_of_birth)`). Yes, it is longer, but it is also more structured, composable, and self-descriptive. These features (which are a direct result of its grammar of graphics lineage) means that both you and your future LLM coding buddy will have an easier time internalizing the workings of *all* types of plots that can be made. The 18 years of dominance of ggplot2 (which shares these features) in the R ecosystem is a testament to this.

As an example, let's change the above plot to instead show the same relationship as a jittered scatterplot.

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter'
```

Or perhaps the jitter follows the distribution of the data so it doubles as a violin plot:

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter', distribution => 'density'
```

> As you can see the syntax and composable nature makes visualization iteration very ergonomic, something that is extremely valuable in both explorative analyses and visualization design.

Reviewer 3: I'd add one example of mapping points to colour (or similar) here.

## Behind the curtains

> With a passing understanding of how to use ggsql, let's discuss what goes on behind the scenes. At its core ggsql is a modular Rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.

Reviewer 3: I'd suggest you drop this entire section and move it to the site somewhere.

---
title: "ggsql: A grammar of graphics for SQL"
date: 2026-04-20
people:
  - Thomas Lin Pedersen
  - Teun Van den Brand
  - George Stagg
  - Hadley Wickham
description: >
  Introducing ggsql, a grammar of graphics for SQL that lets you describe
  visualizations directly inside SQL queries.
image: "ggsql.png"
image-alt: "ggsql logo. A pencil with the body consisting of the standard database silo icon."
topics:
  - Visualization
software:
  - ggsql
languages:
  - SQL
nohero: false
hidesubscription: false
---

Today, we are super excited to announce the alpha-release of [ggsql](https://ggsql.org). As the name suggests, ggsql is an implementation of the grammar of graphics based on SQL syntax, bringing rich, structured visualization support to SQL. It is ready for use in Quarto, Jupyter notebooks, Positron and VS Code among others.

In this post we will go over some of the motivations that lead us to develop this tool, as well as give you ample examples of its use; so you can hopefully get as excited about it as we are.

> ## Meet ggsql

Reviewer: Should add some wording that we have an aperatif of simple plots before the meal of the complete example. The complete example comes out of the blue a bit now.

> That was a lot of code, but it allows us to cover a lot of the most important aspects of the syntax with one example.

Reviewer 2: We haven't seen any code yet.

> That was a lot of code, but it allows us to cover a lot of the most important aspects of the syntax with one example.

Reviewer: Vestigial code leftover from previous section?

### The first plot
To get our feet wet, lets start with the hello-world of visualizations: A scatterplot, using the built-in penguins dataset:

```{ggsql}
VISUALIZE bill_len AS x, bill_dep AS y FROM ggsql:penguins
DRAW point
```

> That wasn't too bad. Sure, it got the verbosity of SQL, but that also means that you can speak out your plot code and understand what it does. We can break down what is going on here line-by-line:

Reviewer 2: ```suggestion
That wasn't too bad. Sure, it has the verbosity of SQL, but that also means that you can speak your plot code out loud and understand what it does. We can break down what is going on here line-by-line:
```

1. We initiate the visual query with `VISUALIZE` and provide a mapping from the built-in penguins dataset, relating `x` to the data in the `bill_len` column, and `y` in the `bill_dep` column.
2. We draw a point layer that, by default, uses the mapping we defined at the top.

With this in place, we can begin to add to the visualization:

```{ggsql}
VISUALIZE bill_len AS x, bill_dep AS y, species AS color FROM ggsql:penguins
DRAW point
```

> We see that a single addition to the mappings add colored categories to the plot. This gradual evolution of plot code is one of the biggest strengths of the grammar of graphics. There are no predefined plot types, only modular parts that can be combined, added, and removed. To further emphasize this, let's add a smooth regression line to the plot:

Reviewer 2: ```suggestion
We see that a single addition to the mappings adds colored categories to the plot. This gradual evolution of plot code is one of the biggest strengths of the grammar of graphics. There are no predefined plot types, only modular parts that can be combined, added, and removed. To further emphasize this, let's add a smooth regression line to the plot:
```

```{ggsql}
VISUALIZE bill_len AS x, bill_dep AS y, species AS color FROM ggsql:penguins
DRAW point
DRAW smooth

>   MAPPING null AS color

Reviewer: I think `MAPPING null AS color` is a hard concept to grasp this early. Perhaps it can make sense to do `DRAW point MAPPING species AS color` in the previous plot. You can then explain local mappings and the regression line then follows without additional complexity (just the concept you can have multi-layer plots). It'd be easier to explain than `null` mappings.

```

> We add a new layer on top of the point layer. This layer also borrows the same mapping as the point layer but overrides the color mapping since we want a single smooth line for all data.

Reviewer 2: It probably doesn't matter in the grand scheme of things, but does adding this kind of smooth line across all the data make sense, in terms of what you'd actually want to visualise?

Maybe we could add `MAPPING species AS stroke` to turn this into a Simpson's paradox example?

After the `DRAW` and `PLACE` clauses we have a `SCALE` clause. This clause controls how data values are translated into values that are meaningful for the aesthetic. In our case, the `category` column holds the strings "Age at mission" and "Age at selection" which doesn't in itself translate to a color value. The clause `SCALE fill TO accent` tells ggsql to use the "accent" color palette when converting the values mapped to fill to actual colors. Scales can be used for much more, like applying transformations to continuous data, defining break points, and setting specific scale types (like ordinal or binned).

The last clause in our visual query is `LABEL` which allows us to add or modify various text labels like title, subtitle, and axis and legend titles. 

### Stepping back
That was a mouthful. But there are two very silvery linings to it all:

1. You now know the most important aspects of the syntax (there are more, of course, but you can grow into that)
2. Many visualization queries will be much simpler than the one above

> We have already seen examples of shorter visual queries above but let's delve on it a bit more with a boxplot of astronaut birth year split by sex:

Reviewer 2: ```suggestion
We have already seen examples of shorter visual queries above but let's continue with a boxplot of astronaut birth year split by sex:
```

The word "delve" is an AI tell, people hate it.

Author: jfc... And I wrote this by hand...

AI spoils everything

* A smaller scope makes it easier to sandbox and prevent malicious code execution (either deliberately or in error)

Both of the above points make ggsql a much more compelling option for integrating into tools such as AI agents assisting you in data analysis, or code based reporting tools that may execute code in different environments.

You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots. This is in stark contrast to most visualization which must first materialize the complete data, then perform the necessary computations on it, then plot it.

### `SELECT pod_door FROM bay WHERE closed`

> LLMs have proven very effective at translating natural language into SQL, and we're bullish that they can be just as effective with ggsql. We've already seen evidence of this in [querychat](https://posit-dev.github.io/querychat/), where you can now visually explore data using natural language via ggsql. And, since ggsql is a much safer and lighter runtime than R or Python, you can much more confidently ship coding agents into a production environment.

Reviewer 2: This might fit better integrated into the paragraph about "integrating into tools such as AI agents assisting you in data analysis". To me it felt a little repetitive here. The feeling is not too strong, though.

Author: Yeah, I couldn't really fit it nicely into the preceding section which is why it ended up here
